src/main/scala/xiangshan/frontend/FTB.scala

```diff
@@ -419,19 +421,126 @@ class FTB(implicit p: Parameters) extends BasePredictor with FTBParams with BPUU
   val btb_enable_dup = dup(RegNext(io.ctrl.btb_enable))
-  val s2_ftb_entry_dup = io.s1_fire.map(f => RegEnable(ftbBank.io.read_resp, f))
+  val s1_read_resp = Mux(s1_close_ftb_req,io.fauftb_entry_in,ftbBank.io.read_resp)
+  val s2_ftbBank_dup = io.s1_fire.map(f => RegEnable(ftbBank.io.read_resp, f))
+  val s2_ftb_entry_dup = dup(0.U.asTypeOf(new FTBEntry))
+  for(((s2_fauftb_entry,s2_ftbBank_entry),s2_ftb_entry) <-
+    s2_fauftb_ftb_entry_dup zip s2_ftbBank_dup zip s2_ftb_entry_dup){
+      s2_ftb_entry := Mux(s2_close_ftb_req,s2_fauftb_entry,s2_ftbBank_entry)
+  }
   val s3_ftb_entry_dup = io.s2_fire.zip(s2_ftb_entry_dup).map {case (f, e) => RegEnable(e, f)}
 
-  val s1_hit = ftbBank.io.read_hits.valid && io.ctrl.btb_enable
-  val s2_hit_dup = io.s1_fire.map(f => RegEnable(s1_hit, 0.B, f))
+
+  //After closing ftb, the hit output from s2 is the hit of FauFTB cached in s1.
+  //s1_hit is the ftbBank hit.
+  val s1_hit = Mux(s1_close_ftb_req,false.B,ftbBank.io.read_hits.valid && io.ctrl.btb_enable)
+  val s2_ftb_hit_dup = io.s1_fire.map(f => RegEnable(s1_hit, 0.B, f))
+  val s2_hit_dup = dup(0.U.asTypeOf(Bool()))
+  for(((s2_fauftb_hit,s2_ftb_hit),s2_hit) <-
+    s2_fauftb_ftb_entry_hit_dup zip s2_ftb_hit_dup zip s2_hit_dup){
+      s2_hit := Mux(s2_close_ftb_req,s2_fauftb_hit,s2_ftb_hit)
+  }
   val s3_hit_dup = io.s2_fire.zip(s2_hit_dup).map {case (f, h) => RegEnable(h, 0.B, f)}
-  val writeWay = ftbBank.io.read_hits.bits
+  val writeWay = Mux(s1_close_ftb_req,0.U,ftbBank.io.read_hits.bits)
+
+  def ftbslot_compare(x: FtbSlot, y: FtbSlot) = {
+    VecInit(
+      x.offset  === y.offset,
+      x.lower   === y.lower,
+      x.tarStat === y.tarStat,
+      x.sharing === y.sharing,
+      x.valid   === y.valid
+    )
+  }
+
+  def ftbentry_compare(x: FTBEntry, y: FTBEntry) = {
+    val validDiff     = x.valid === y.valid
+    val brSlotsDiffVec  : IndexedSeq[Vec[Bool]] =
+      x.brSlots.zip(y.brSlots).map{
+        case(xSlot,ySlot) => ftbslot_compare(xSlot,ySlot)
+      }
+    val brSlotsDiffSeq  : IndexedSeq[Bool] = brSlotsDiffVec.map(_.reduce(_&&_))
+    val tailSlotDiffSeq : IndexedSeq[Bool] = ftbslot_compare(x.tailSlot,y.tailSlot).toIndexedSeq
+    val pftAddrDiff   = x.pftAddr  === y.pftAddr
+    val carryDiff     = x.carry  === y.carry
+    val isCallDiff    = x.isCall === y.isCall
+    val isRetDiff     = x.isRet  === y.isRet
+    val isJalrDiff    = x.isJalr === y.isJalr
+    val lastMayBeRviCallDiff = x.last_may_be_rvi_call === y.last_may_be_rvi_call
+    val alwaysTakenDiff : IndexedSeq[Bool] =
+      x.always_taken.zip(y.always_taken).map{
+        case(xAT,yAT) => xAT === yAT
+      }
+    VecInit(
+      validDiff,
+      brSlotsDiffSeq.reduce(_&&_),
+      tailSlotDiffSeq.reduce(_&&_),
+      pftAddrDiff,
+      carryDiff,
+      isCallDiff,
+      isRetDiff,
+      isJalrDiff,
+      lastMayBeRviCallDiff,
+      alwaysTakenDiff.reduce(_&&_)
+    )
+  }
```

Comment:
Maybe we can move these two functions to FtbSlot class and FTBEntry class

Reply:
ok